noir_stdlib/src/meta/map.nr

```diff
@@ -0,0 +1,459 @@
+use crate::cmp::Eq;
+use crate::collections::vec::Vec;
+use crate::option::Option;
+use crate::default::Default;
+use crate::hash::{Hash, Hasher, BuildHasher, BuildHasherDefault};
+use crate::hash::poseidon2::Poseidon2;
+use crate::collections::bounded_vec::BoundedVec;
+
+// Compile-time Hash table with open addressing and quadratic probing.
+struct CtHashMap<K, V, B> {
+    _table: [Slot<K, V>],
+
+    // Amount of valid elements in the map.
+    _len: u32,
+
+    _build_hasher: B
+}
+
+// Data unit in the CtHashMap table.
+// In case Noir adds support for enums in the future, this  
+// should be refactored to have three states:
+// 1. (key, value)
+// 2. (empty)
+// 3. (deleted)
+struct Slot<K, V> {
+    _key_value: Option<(K, V)>,
```

Comment:
What is the difference between a `_key_value` of `None` and one with `Some` whose `_is_deleted` field is `true`?

I saw the following comment on `mark_deleted` and am not sure why they need to be differentiated for lookup. Why not return `None` from `get` both when empty and deleted?

```rust
    // Shall not override `_key_value` with Option::none(),
    // because we must be able to differentiate empty 
    // and deleted slots for lookup.
```

Reply:
`_is_deleted` is the tombstone value which is standard for hashmaps with open addressing (store elements in the hashmap directly rather than in separate buckets). It is required to keep `get` working when hash collisions are possible. E.g. consider a hashmap where we insert there elements with the same hash: `[..., (1, 1), (2, 2), (3, 3), ...]`.

When calling get we perform `let index = elem.hash() % self.capacity()` then check the key of that element to ensure it is matching. If the key doesn't match, we check the next element in the hashmap and so on until we come to an empty slot in the hashmap or a slot where the hash of that key no longer matches our hash.

With that in mind lets say we `remove` the `2` key: `[..., (1, 1), _, (3, 3), ...]`. Not if we wanted to call `get` to retrieve the `3` the `get` call would fail since it would start at `(1, 1)`, find the empty slot `2` left behind, and stop. This is why when we remove elements we mark them with a tombstone so that we can check this flag during `get` and if so, continue on to check the elements after as well. That's why we need to distinguish between empty slots and tombstones.